File: services/control-plane/app/domain/permissions.py
"""V1 role -> permission matrix.

Administrator: full V1 access ("*").
Operator: deploy/scale/restart/stop/rollback/view — no user/security admin.
Viewer: read-only.
"""

ROLE_PERMISSIONS: dict[str, set[str]] = {
    "Administrator": {"*"},
    "Operator": {"view", "deploy", "scale", "restart", "stop", "rollback"},
    "Viewer": {"view"},
}


def has_permission(roles: set[str], permission: str) -> bool:
    for role in roles:
        granted = ROLE_PERMISSIONS.get(role, set())
        if "*" in granted or permission in granted:
            return True
    return False

File: apps/dashboard/src/lib/permissions.ts
// Mirrors services/control-plane/app/domain/permissions.py's role -> permission
// matrix, so the dashboard can hide destructive/administrative actions from
// users who don't hold them — a UX courtesy only. The Control Plane API is
// the actual enforcement point; every mutating request is re-checked there
// regardless of what this file decides to show or hide.
const ROLE_PERMISSIONS: Record<string, Set<string> | "*"> = {
  Administrator: "*",
  Operator: new Set(["view", "deploy", "scale", "restart", "stop", "rollback"]),
  Viewer: new Set(["view"]),
};

export function hasPermission(roles: string[], permission: string): boolean {
  for (const role of roles) {
    const granted = ROLE_PERMISSIONS[role];
    if (granted === "*") return true;
    if (granted?.has(permission)) return true;
  }
  return false;
}

export function isAdministrator(roles: string[]): boolean {
  return roles.includes("Administrator");
}
